ddtrace/internal/_buffer.pyx

```diff
@@ -0,0 +1,80 @@
+from collections import deque
+import threading
+
+from ddtrace.vendor import attr
+from ..internal.logger import get_logger
+from . import _rand
+
+
+log = get_logger(__name__)
+
+
+@attr.s
+class TraceBuffer(object):
+
+    maxsize = attr.ib(type=int, default=0)
+    _size = attr.ib(init=False, type=int, default=0, repr=True)
+    _lock = attr.ib(init=False, factory=threading.Lock, repr=False)
+    _buffer = attr.ib(init=False, factory=deque, repr=False)
+    _accepted = attr.ib(init=False, type=int, default=0)
+    _accepted_lengths = attr.ib(init=False, type=int, default=0)
+    _dropped = attr.ib(init=False, type=int, default=0)
+
+    def __len__(self):
+        return len(self._buffer)
+
+    @property
+    def size(self):
+        with self._lock:
+            return self._size
+
+    def clear(self):
+        self._buffer.clear()
+        self._size = 0
+
+    def put(self, item):
+        with self._lock:
+            if self.maxsize <= 0 or self._size + len(item) < self.maxsize:
+                self._buffer.append(item)
+            else:
+                self._dropped += 1
+                self._dropped_size += len(item)
+                log.warning("trace buffer %r is full, dropping a random trace", self)
+
+            self._accepted += 1
+            self._accepted_lengths += len(item)
+
+    def popleft(self):
+        i = self._buffer.popleft()
+        self._size -= len(i)
+        return i
+
+    def get(self, size=None):
```

Comment:
I wonder if we can just do:

```python
def get():
    self._buffer, buffer = deque(), self._buffer
    return self._buffer
```

wonder what impact that has on GC, cleaning up an entire `deque` instance vs clearing everyone out of the existing object?